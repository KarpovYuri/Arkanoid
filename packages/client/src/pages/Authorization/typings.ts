import { StringObject } from '../../typings';

export type LoginProps = {
  onLogin(userData: StringObject): void;
};
