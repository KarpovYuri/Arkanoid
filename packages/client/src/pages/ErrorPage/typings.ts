export type ErrorPageProps = {
  message?: string;
  status?: number;
};
