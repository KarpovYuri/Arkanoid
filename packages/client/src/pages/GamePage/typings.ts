export enum GameStatus {
  ONBOARDING = 'onboarding',
  PREPARING = 'preparing',
  RUNNING = 'running',
  WIN = 'win',
  LOSE = 'lose',
}
