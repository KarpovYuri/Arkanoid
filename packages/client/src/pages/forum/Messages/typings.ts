export type TMessage = {
  id: number
  user_id: number
  createdAt: string
  content: string
  forum_id: number
  parent_id?: number
  user?: any
};

export type TMessageNew = {
  content: string
  user_id: number
  topic_id: number
  forum_id: number
  parent_id?: number
  author?: string
};
