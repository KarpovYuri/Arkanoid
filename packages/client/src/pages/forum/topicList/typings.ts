export type TTopic = {
  id: number | string;
  name: string;
  countAnswer: number;
  lastMessageTime: string;
};
