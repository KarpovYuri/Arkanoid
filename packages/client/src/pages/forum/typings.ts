export type TForum = {
  id: number
  name: string
  countTopic: number
  countAnswer: number
};
