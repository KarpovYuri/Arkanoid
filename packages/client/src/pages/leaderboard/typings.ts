export enum Order {
  ASC = 'asc',
  DESC = 'desc',
}

export type TLeaderBoard = {
  id: number;
  name: string;
  avatar: string;
  points: number;
  place: number;
};

export type TLeaderBoardProps = {
  data: TLeaderBoard;
  key: number;
  iam: boolean;
};

export type SortEventProps = {
  sorting: (sortField: keyof TLeaderBoard, order: Order) => void;
};

export type Sorting = {
  sort: keyof TLeaderBoard;
  order: Order;
};
