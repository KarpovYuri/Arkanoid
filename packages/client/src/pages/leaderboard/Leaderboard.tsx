import React, { useEffect, useState } from 'react';
import { leaderboardAPI } from '../../api/LeaderboardAPI/LeaderboardAPI';
import { TableHead } from '../../components/Leaderboard/TableHead';
import { TableRow } from '../../components/Leaderboard/TableRow';
import { Loader } from '../../components/Loader';
import styles from './Leaderboard.module.css';
import { Order, TLeaderBoard } from './typings';

export function Leaderboard() {
  const [leaders, setLeaders] = useState<TLeaderBoard[]>([
    {
      id: 123,
      name: 'Безумный Майк',
      score: 999999,
      place: 1,
    },
    {
      id: 1234,
      name: 'Вася Петров',
      score: 44234,
      place: 2,
    },
    {
      id: 12345,
      name: 'Аня Иванов',
      score: 23322,
      place: 3,
    },
    {
      id: 123456,
      name: 'Владимир Ильич',
      score: 20000,
      place: 4,
    },
    {
      id: 3213,
      name: 'Кот',
      avatar: '/avatar.jpg',
      score: 12,
      place: 5,
    },
    {
      id: 12123,
      name: 'Кот',
      avatar: '/avatar.jpg',
      score: 12,
      place: 6,
    },
    {
      id: 32,
      name: 'Кот',
      avatar: '/avatar.jpg',
      score: 12,
      place: 7,
    },
    {
      id: 1233,
      name: 'Кот',
      avatar: '/avatar.jpg',
      score: 12,
      place: 8,
    },
    {
      id: 33,
      name: 'Кот',
      avatar: '/avatar.jpg',
      score: 12,
      place: 9,
    },
  ]);
  const [loading, setLoading] = useState(true);

  function fetchLeaders() {
    const promise = leaderboardAPI.getAllLiders({
      ratingFieldName: 'score',
      cursor: 0,
      limit: 15,
    });

    promise
      .then((result) => {
        setLeaders(result.map((item: Record<string, TLeaderBoard>, index: number) => {
          item.data.place = index + 1;
          if (!item.data.name) {
            item.data.name = 'No name';
          }
          return item.data;
        }));
      })
      .catch((error) => {
        throw Error(`Something wrong with Leaderboard - ${error}`);
      })
      .finally(() => {
        setLoading(false);
      });
  }

  useEffect(() => {
    fetchLeaders();
  }, []);

  const sortLeaders = (sortField: keyof Omit<TLeaderBoard, 'avatar'>, orders: Order): void => {
    if (sortField !== null) {
      setLeaders([...leaders].sort((a, b) => {
        if (a[sortField] < b[sortField]) {
          return orders === 'desc' ? -1 : 1;
        }
        if (a[sortField] > b[sortField]) {
          return orders === 'desc' ? 1 : -1;
        }
        return 0;
      }));
    }
  };

  /**
   * id текущего игрока - пока захардкодено
   * TODO: вытащить id текущего пользователя
   */
  const currentPlayerId = 3;

  return (
    <main className={styles.leaderboard}>
      <section className={styles.wrapper}>
        <h1 className={styles.title}>Рейтинг игроков</h1>

        {loading
          ? <Loader />
          : (
            <table className={styles.table}>
              <TableHead sorting={sortLeaders} />
              <tbody>
                {leaders.map((leader, index) => (
                  <TableRow row={{
                    data: leader,
                    key: leader.id,
                    iam: (index === currentPlayerId),
                  }}
                  />
                ))}
              </tbody>
            </table>
          )}

      </section>
    </main>
  );
}
