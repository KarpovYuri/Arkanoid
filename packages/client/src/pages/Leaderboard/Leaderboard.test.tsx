import { act, render, screen, waitFor } from '@testing-library/react';
import { enableFetchMocks } from 'jest-fetch-mock';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { AnyAction, configureStore, Store } from '@reduxjs/toolkit';
import { rootReducer } from '../../store/store';
import { TLeaderBoard } from './typings';
import { BASE_URL_YANDEX, LEADERBOARD, TEAM_NAME } from '../../utils/apiConstans';
import { useAppSelector } from '../../store/hooks';
import { useLeaders } from '../../hooks/useLeaders';
import { Leaderboard } from './Leaderboard';

enableFetchMocks();

jest.mock('../../store/hooks');
jest.mock('../../hooks/useLeaders', () => ({
  useLeaders: jest.fn(),
}));

const mockData: TLeaderBoard[] = [
  { id: 1, name: 'User1', score: 200, place: 1, avatar: '/ava1.png' },
  { id: 2, name: 'User2', score: 100, place: 2, avatar: '/ava2.png' },
];

describe('Leaderboard', () => {
  let store: Store<unknown, AnyAction>;
  const originalUseLeaders = jest.requireActual('../../hooks/useLeaders').useLeaders;

  beforeEach(() => {
    fetchMock.resetMocks();
    store = configureStore({ reducer: rootReducer });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  beforeAll(() => {
    (useLeaders as jest.Mock).mockImplementation(originalUseLeaders);
  });

  test('it renders with data and return leader rows', async () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));

    // Мокаем useAppSelector для возврата фиктивного userData
    (useAppSelector as jest.Mock).mockReturnValue({ id: 2 });

    // Мокаем useLeaders для возврата фиктивной функции и данных для isLoading и leadersError
    (useLeaders as jest.Mock).mockImplementation((setLeaders) => {
      const fetchLeaders = () => {
        setLeaders(mockData);
      };

      return [fetchLeaders, false, null];
    });

    render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>,
    );

    const leaderboard = screen.getByTestId('leaderboard');

    expect(leaderboard).toBeInTheDocument();
    expect(leaderboard).toMatchSnapshot();

    // Ожидание обновления состояния после выполнения useEffect и fetch-запроса
    await waitFor(() => screen.getByTestId('leaderboard-table'));

    // Проверяем, что отрисовывается таблица с данными
    expect(screen.getByTestId('leaderboard-table')).toBeInTheDocument();

    // Проверяем, что отрисовывается TableRow для каждого элемента данных
    mockData.forEach((leader) => {
      expect(screen.getByText(leader.name)).toBeInTheDocument();
      expect(screen.getByText(leader.score.toString())).toBeInTheDocument();
    });
  });

  test('it renders correctly without data', () => {
    // Восстанавливаем оригинальный хук useLeaders
    (useLeaders as jest.Mock).mockImplementation(originalUseLeaders);

    act(() => render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>,
    ));

    const leaderboard = screen.getByTestId('leaderboard');

    expect(leaderboard).toBeInTheDocument();
    expect(leaderboard).toMatchSnapshot();
  });

  test('loader is shown', () => {
    // Восстанавливаем оригинальный хук useLeaders
    (useLeaders as jest.Mock).mockImplementation(originalUseLeaders);

    act(() => render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>,
    ));

    const loader = screen.getByTestId('loader');

    expect(loader).toBeInTheDocument();
  });

  test('it sends a fetch to correct url', () => {
    fetchMock.mockResponseOnce(JSON.stringify(mockData));

    // Восстанавливаем оригинальный хук useLeaders
    (useLeaders as jest.Mock).mockImplementation(originalUseLeaders);

    act(() => render(
      <Provider store={store}>
        <Leaderboard />
      </Provider>,
    ));

    // Ожидаем, что fetch будет вызван 1 раз
    expect(fetch).toHaveBeenCalledTimes(1);

    const [requestUrl] = fetchMock.mock.calls[0];

    // Ожидаем, что fetch вызван с верным урлом Яндекса
    expect(requestUrl).toBe(`${BASE_URL_YANDEX}${LEADERBOARD}/${TEAM_NAME}`);
  });
});
