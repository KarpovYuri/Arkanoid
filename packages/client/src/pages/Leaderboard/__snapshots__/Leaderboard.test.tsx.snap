// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard it renders correctly without data 1`] = `
<main
  class="leaderboard"
  data-testid="leaderboard"
>
  <section
    class="wrapper"
  >
    <h1
      class="title"
    >
      Рейтинг игроков
    </h1>
    <div
      class="loader"
      data-testid="loader"
    >
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
      <div />
    </div>
  </section>
</main>
`;
